### DISCUSSION TITLE
Query parameters of type `list` are recognized as body parameters, when nested inside a pydantic model.

### DISCUSSION THREAD

**Discussion by nobletooth** [fastapi/fastapi#10556] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Query, Depends
from pydantic import BaseModel


class Request(BaseModel):
    scalar_parameter: int = Query()  # perfectly fine, added to parameters.
    list_parameter: list[int] = Query()  # this is moved to body instead of query.


app = FastAPI()


@app.get(path="/resource")
def get_resource(req: Request = Depends()) -> str:
    return "resource"
```

### Description

* Open the browser, look as /docs endpoint to see swagger docs.
* I expected to see `list_parameter` in the query parameters like `scalar_parameter`
* I see `list_parameter` in the request body, which is not even usable as request bodies are actually prohibited in `GET` requests.

### Operating System

macOS

### FastAPI Version

0.104.1

### Pydantic Version

2.4.2

### Python Version

3.12.0

### Additional Context

I know a workaround that I'll paste here. My codebase however has lots of similar requests that require the same boilerplate code which is a pain.

```python
from fastapi import FastAPI, Query, Depends
from pydantic import BaseModel


class Request(BaseModel):
    scalar_parameter: int = Query()  # perfectly fine, added to parameters.
    list_parameter: list[int]  # hacked

    def __init__(self, list_parameter: list[int] = Query(), **kwargs) -> None:
        super().__init__(list_parameter=list_parameter, **kwargs)


app = FastAPI()


@app.get(path="/resource")
def get_resource(req: Request = Depends()) -> str:
    return "resource"
```

**Comment by White-Mask** [comment, marked as answer]
Hello @nobletooth, the issue you are experiencing is due to the way FastAPI handles Pydantic models for query parameters. According to the FastAPI documentation and several issue reports on GitHub, FastAPI has limitations in handling Pydantic models for query parameters, especially when the parameters are lists or complex types.

In GET operations, you cannot send a request body in the HTTP request. Therefore, when using a Pydantic model for query parameters, FastAPI attempts to interpret the parameters as a JSON body, which is not valid for GET operations. This is why list-type parameters appear in the request body instead of being treated as query parameters.

In your case, you are using a Pydantic model for query parameters, which is causing the issue. FastAPI cannot correctly interpret list-type parameters as query parameters, which is why it displays them in the request body instead of as query parameters.

I know that Pydantic can be very useful, but as I mentioned earlier, this is an issue that has been under discussion for some time.
Now, I will show you options on how to address this issue. https://github.com/tiangolo/fastapi/issues/2869 https://github.com/tiangolo/fastapi/issues/318

As you already know, the FastAPI documentation suggests naming each parameter in the function like this:

```python
@app.get("/resource")
async def get_resource(scalar_parameter: int = Query(...), list_parameter: List[int] = Query(None)) -> str:
    return "resource" 
```

Another method mentioned in the documentation, which you've already mentioned, is to do it this way:
```python
from fastapi import FastAPI, Depends, Query

app = FastAPI()

class Request:
    def __init__(
        self,
        scalar_parameter: int = Query(None),
        list_parameter: list[int] = Query(None)
    ):
        self.scalar_parameter = scalar_parameter
        self.list_parameter = list_parameter


@app.get("/resource")
def search(req: Request = Depends()) -> str:
    return "resource"
```
* https://github.com/tiangolo/fastapi/issues/318#issuecomment-507043221
* https://fastapi.tiangolo.com/tutorial/dependencies/classes-as-dependencies/#classes-as-dependencies_1

Another way to address your issue is like this:
```python
from fastapi import FastAPI, Depends, Query
from dataclasses import dataclass

app = FastAPI()

@dataclass
class Request:
    scalar_parameter: int = Query(None)
    list_parameter: list[int] = Query(None)

@app.get("/resource")
async def get_resource(req: Request = Depends()) -> str:
  return "resource"
```
I got this solution from a past discussion in the forum
* https://gist.github.com/Atheuz/075f4d8fe3b56d034741301ba2574ef1
* https://github.com/tiangolo/fastapi/issues/318#issuecomment-506989873

These are the solutions I found and have been testing. I know they don't completely resolve your issue (as I mentioned, it's a long-standing error), but I hope this has helped you with your problem. Best regards.

**Reply by nobletooth** [reply]
Thanks for you informative reply. I was trying to find / implement a hack but I couldn't. This was useful tho.

**Comment by loucadufault** [comment]
These caveats should be documented in https://fastapi.tiangolo.com/tutorial/query-param-models/
